cpp/src/strings/convert/convert_urls.cu

```diff
@@ -0,0 +1,273 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/strings/convert/convert_urls.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <strings/utilities.hpp>
+#include <strings/utilities.cuh>
+
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+namespace
+{
+
+//
+// This is the functor for the url_encode() method below.
+// Specific requirements are documented in custrings issue #321.
+// In summary it converts mostly non-ascii characters and control characters into UTF-8 hex characters
+// prefixed with '%'. For example, the space character must be converted to characters '%20' where the
+// '20' indicates the hex value for space in UTF-8. Likewise, multi-byte characters are converted to
+// multiple hex charactes. For example, the é character is converted to characters '%C3%A9' where 'C3A9'
+// is the UTF-8 bytes xc3a9 for this character.
+//
+struct url_encoder_fn
+{
+    column_device_view const d_strings;
+    int32_t const* d_offsets{};
+    char* d_chars{};
+
+    // utility to create 2-byte hex characters from single binary byte
+    __device__ void byte_to_hex( u_char byte, char* hex )
+    {
+        hex[0] = '0';
+        if( byte >= 16 )
+        {
+            u_char hibyte = byte/16;
+            hex[0] = hibyte < 10 ? '0'+hibyte : 'A'+(hibyte-10);
+            byte = byte - (hibyte * 16);
+        }
+        hex[1] = byte < 10 ? '0'+byte : 'A'+(byte-10);
+    }
+
+    __device__ bool should_not_url_encode( char ch )
+    {
+        return ( (ch>='0' && ch<='9') || // these are the characters
+                 (ch>='A' && ch<='Z') || // that are not to be url encoded
+                 (ch>='a' && ch<='z') || // reference: docs.python.org/3/library/urllib.parse.html#urllib.parse.quote
+                 (ch=='.') || (ch=='_') || (ch=='~') || (ch=='-') );
+    }
+
+    // main part of the functor the performs the url-encoding
+    __device__ size_type operator()( size_type idx )
+    {
+        if( d_strings.is_null(idx) )
+            return 0;
+        string_view d_str = d_strings.element<string_view>(idx);
+        //
+        char* out_ptr = d_chars ? d_chars + d_offsets[idx] : nullptr;
+        size_type nbytes = 0;
+        char hex[2]; // two-byte hex max
+        for( auto itr = d_str.begin(); itr!=d_str.end(); ++itr )
+        {
+            auto ch = *itr;
```

Comment:
Range based for:
```cpp
        for (auto const ch : d_str) {
```

Reply:
I looked at that. But I need the iterator at line 90:
https://github.com/rapidsai/cudf/blob/5b0722151136f694da37beae8c31e63e03ec8be2/cpp/src/strings/convert/convert_urls.cu#L90

Reply:
My bad, missed that

Reply:
Seems to me the thing iterated over could factor out the interface used here when iterating. E.g. rather than `byte_offset()` being a method of the iterator, make it a method of the thing iterated over. Maybe that thing is a std::pair, or maybe it's an intermediate struct. I would guess this would clean up a lot of algorithms in the strings code, and it would make the iterator more standard -- which would make it more composable, e.g. with fancy iterators.

Reply:
Getting the byte-offset during iteration is very useful since it does not require the expense of computing the byte-offset.
I think creating an iterator that returns a pair which includes the character and the offset so we can do more algorithmic things would be a good enhancement.